s3_start_transfer/src/s3_start_transfer.py

```diff
@@ -0,0 +1,206 @@
+    :param data: Dict of data to post
+    :returns: dict of json data returned by request
+    """
+    am_url = os.environ["ARCHIVEMATICA_URL"]
+    am_user = os.environ["ARCHIVEMATICA_USERNAME"]
+    am_api_key = os.environ["ARCHIVEMATICA_API_KEY"]
+    am_headers = {"Authorization": "ApiKey {0}:{1}".format(am_user, am_api_key)}
+
+    url = "{0}{1}".format(am_url, url)
+    print("URL: %s; Data: %s" % (url, data))
+    response = requests.post(url, json=data, headers=am_headers)
+    print("Response: %s" % response)
+    response_json = response.json()
+    print("Response JSON: %s" % response_json)
+    return response_json
+
+
+def ss_api_get(api_path, params=None):
+    """
+    GET request to the Archivematica storage service API
+    :param api_path: URL path to request (without hostname, e.g. /api/v2/location/)
+    :param params: Dict of params to include in the request
+    :returns: dict of json data returned by request
+    """
+    ss_url = os.environ["ARCHIVEMATICA_SS_URL"]
+    ss_user = os.environ["ARCHIVEMATICA_SS_USERNAME"]
+    ss_api_key = os.environ["ARCHIVEMATICA_SS_API_KEY"]
+    ss_headers = {"Authorization": "ApiKey {0}:{1}".format(ss_user, ss_api_key)}
+
+    params = params or {}
+    url = "{0}{1}".format(ss_url, api_path)
+    print("URL: %s; Params: %s" % (url, params))
+    response = requests.get(url, params=params, headers=ss_headers)
+    print("Response: %s" % response)
+    response_json = response.json()
+    print("Response JSON: %s" % response_json)
+    return response_json
+
+
+def get_target_path(bucket, key):
+    """
+    Get the uploaded file's path on the Archivematica storage service,
+    as a `<location_uuid>:<target_path>` bytestring returned by `find_matching_path`
+
+    :param bucket: Name of s3 bucket
+    :param key: Name of s3 key
+    :returns: bytestring identifying the path
+    """
+
+    # Get s3 location matching bucket and key
+    # Look for an S3 transfer source that matches the bucket name and prefix
+    # and take the transfer ID of that location
+    """
+    Example of relevant fields in Location API return json:
+    {
+        "meta": ...
+        "objects": [{
+            "description": "S3 Transfer Source",
+            "purpose": "TS",
+            "relative_path": "/test-uploads/",
+            "space": "/api/v2/space/6710c8dd-00ad-4614-8f1c-d9be23052179/",
+            "uuid": "017fcad6-fb5c-434e-818a-14b812ef6427"
+            ...
+            }]
+    }
+    """
+    s3_sources = ss_api_get(
+        "/api/v2/location/", {"space__access_protocol": "S3", "purpose": "TS"}
+    )
+
+    """
+    Example of relevant fields in Space API json:
+    {
+        "access_protocol": "S3",
+        "s3_bucket": "wellcomecollection-archivematica-transfer-source",
+        "uuid": "6710c8dd-00ad-4614-8f1c-d9be23052179",
+        ...
+    }
+    """
+    all_spaces = OrderedDict(
+        (location["space"], None) for location in s3_sources["objects"]
+    )
```

Comment:
Not in actually running the code - the instability of set ordering does vary the order in which the API calls are made, which makes the tests unstable. I'm not super happy about it but now I think about it some more I suppose it's better to make the tests a little more complex rather than the code...?

Reply:
I don’t mind where the complexity lives especially; if we leave it here it’d be good to have an explanation of why we're using an unusual data structure here.